File: src/content/docs/topics/cs/design-patterns/command.go
package main

import "fmt"

type Command interface {
	Execute()
	Undo()
}

type TextEditor struct {
	content string
}

func (e *TextEditor) GetContent() string { return e.content }

func (e *TextEditor) Insert(text string, pos int) {
	e.content = e.content[:pos] + text + e.content[pos:]
}

func (e *TextEditor) Delete(pos, length int) {
	e.content = e.content[:pos] + e.content[pos+length:]
}

type InsertCommand struct {
	editor   *TextEditor
	text     string
	position int
}

func (c *InsertCommand) Execute() { c.editor.Insert(c.text, c.position) }
func (c *InsertCommand) Undo()    { c.editor.Delete(c.position, len(c.text)) }

type DeleteCommand struct {
	editor   *TextEditor
	position int
	length   int
	deleted  string
}

func (c *DeleteCommand) Execute() {
	content := c.editor.GetContent()
	c.deleted = content[c.position : c.position+c.length]
	c.editor.Delete(c.position, c.length)
}

func (c *DeleteCommand) Undo() { c.editor.Insert(c.deleted, c.position) }

type CommandHistory struct {
	history []Command
	cursor  int
}

func NewHistory() *CommandHistory { return &CommandHistory{cursor: -1} }

func (h *CommandHistory) Execute(cmd Command) {
	h.history = h.history[:h.cursor+1]
	cmd.Execute()
	h.history = append(h.history, cmd)
	h.cursor++
}

func (h *CommandHistory) Undo() {
	if h.cursor < 0 {
		return
	}
	h.history[h.cursor].Undo()
	h.cursor--
}

func (h *CommandHistory) Redo() {
	if h.cursor >= len(h.history)-1 {
		return
	}
	h.cursor++
	h.history[h.cursor].Execute()
}

func main() {
	editor := &TextEditor{}
	history := NewHistory()

	history.Execute(&InsertCommand{editor, "Hello, ", 0})
	history.Execute(&InsertCommand{editor, "world", 7})
	fmt.Println(editor.GetContent()) // Hello, world

	history.Undo()
	fmt.Println(editor.GetContent()) // Hello,

	history.Redo()
	fmt.Println(editor.GetContent()) // Hello, world
}

File: src/content/docs/topics/cs/design-patterns/command.ts
interface Command {
  execute(): void;
  undo(): void;
}

class TextEditor {
  private content = '';

  getContent(): string { return this.content; }

  insert(text: string, pos: number): void {
    this.content = this.content.slice(0, pos) + text + this.content.slice(pos);
  }

  delete(pos: number, len: number): void {
    this.content = this.content.slice(0, pos) + this.content.slice(pos + len);
  }
}

class InsertCommand implements Command {
  constructor(
    private editor: TextEditor,
    private text: string,
    private position: number,
  ) {}

  execute(): void { this.editor.insert(this.text, this.position); }
  undo(): void { this.editor.delete(this.position, this.text.length); }
}

class DeleteCommand implements Command {
  private deleted = '';

  constructor(
    private editor: TextEditor,
    private position: number,
    private length: number,
  ) {}

  execute(): void {
    const content = this.editor.getContent();
    this.deleted = content.slice(this.position, this.position + this.length);
    this.editor.delete(this.position, this.length);
  }

  undo(): void { this.editor.insert(this.deleted, this.position); }
}

class CommandHistory {
  private history: Command[] = [];
  private cursor = -1;

  execute(cmd: Command): void {
    this.history = this.history.slice(0, this.cursor + 1);
    cmd.execute();
    this.history.push(cmd);
    this.cursor++;
  }

  undo(): void {
    if (this.cursor < 0) return;
    this.history[this.cursor].undo();
    this.cursor--;
  }

  redo(): void {
    if (this.cursor >= this.history.length - 1) return;
    this.cursor++;
    this.history[this.cursor].execute();
  }
}

const editor = new TextEditor();
const history = new CommandHistory();

history.execute(new InsertCommand(editor, 'Hello, ', 0));
history.execute(new InsertCommand(editor, 'world', 7));
console.log(editor.getContent()); // Hello, world

history.undo();
console.log(editor.getContent()); // Hello,

history.redo();
console.log(editor.getContent()); // Hello, world

history.execute(new DeleteCommand(editor, 7, 5));
console.log(editor.getContent()); // Hello,
